aggregator/src/datastore.rs

```diff
@@ -2885,6 +3070,155 @@ ORDER BY id DESC
+        self.tx
+            .execute(
+                &stmt,
+                &[
+                    /* task_id */ &task_id.get_encoded(),
+                    /* client_timestamp */
+                    &oldest_allowed_report_timestamp.as_naive_date_time()?,
+                ],
+            )
+            .await?;
+        Ok(())
+    }
+
+    /// Deletes old aggregation artifacts (aggregation jobs/report aggregations/batch aggregations)
+    /// for a given task, that is, aggregation artifacts for which all associated client reports
+    /// have timestamps older than a given timestamp, and which are not included in any collection
+    /// artifacts.
+    ///
+    /// Only implemented for leader tasks with the time-interval query type.
+    ///
+    /// After calling this function, delete_old_client_reports must be called in the same
+    /// transaction to avoid re-aggregating client reports.
+    #[tracing::instrument(skip(self), err)]
+    pub async fn delete_old_aggregation_artifacts(
+        &self,
+        task_id: &TaskId,
+        oldest_allowed_report_timestamp: Time,
+    ) -> Result<(), Error> {
+        let stmt = self
+            .tx
+            .prepare_cached(
+                "WITH aggregation_jobs_to_delete AS (
+                    SELECT id FROM aggregation_jobs
+                    JOIN (
+                        SELECT
+                            report_aggregations.aggregation_job_id,
+                            MAX(client_reports.client_timestamp) AS max_timestamp
+                        FROM report_aggregations
+                        JOIN client_reports
+                            ON client_reports.id = report_aggregations.client_report_id
+                        GROUP BY report_aggregations.aggregation_job_id
+                    ) report_max_timestamps
+                        ON report_max_timestamps.aggregation_job_id = aggregation_jobs.id
+                    WHERE aggregation_jobs.task_id = (SELECT id FROM tasks WHERE task_id = $1)
+                      AND report_max_timestamps.max_timestamp < $2
+                      AND NOT EXISTS (
+                          SELECT id FROM collect_jobs
+                          WHERE aggregation_jobs.task_id = collect_jobs.task_id
+                            AND aggregation_jobs.batch_interval <@ collect_jobs.batch_interval)
+                ),
+                deleted_report_aggregations AS (
+                    DELETE FROM report_aggregations
+                    WHERE report_aggregations.aggregation_job_id IN (
+                        SELECT id FROM aggregation_jobs_to_delete)
+                ),
+                deleted_batch_aggregations AS (
+                    DELETE FROM batch_aggregations
+                    WHERE batch_aggregations.batch_identifier IN (
+                        SELECT batch_identifier FROM aggregation_jobs
+                        WHERE aggregation_jobs.id IN (SELECT id FROM aggregation_jobs_to_delete))
+                )
```

Comment:
Neither `deleted_report_aggregations` or `deleted_batch_aggregations` are used elsewhere in the query. Would there be a downside to running these two statements as separate calls to `self.tx.prepare_cached` and `self.tx.execute` in this method, run after deleting the aggregation jobs? I think it would be a little more clear and we should get the same transaction semantics, right?

Or perhaps doing all of this in one query allows Postgres to run it more efficiently?

Reply:
I think the only way this is more efficient is in that it takes only one network round-trip rather than three. This is probably marginal, but I can't think of a good reason to split it out once it's written this way -- we will always want to fuse these operations together.